### PR TITLE
Enhancements to the SMS User Registration Workflow

### PR DESCRIPTION
**Pull request by gcapalbo**
https://docs.google.com/document/d/16i2wZzXjvT1lJEke8mL10fJPQJsSDkI0yh7g5XZ7qEs/edit#

Adds the following features to the SMS user registration workflow:
- adds email to the signup page
- adds an option to assume the recipient phones are android without asking them
- adds an option for a custom registration message
- exposes an api to initiate the workflow

Best reviewed commit by commit

### CODE REVIEW

File: corehq/apps/sms/resources/v0_5.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
+import json
+from collections import namedtuple
+from corehq.apps.accounting.utils import domain_has_privilege
+from corehq.apps.api.resources import HqBaseResource
+from corehq.apps.api.resources.v0_1 import RequirePermissionAuthentication
+from corehq.apps.app_manager.dbaccessors import get_app
+from corehq.apps.domain.models import Domain
+from corehq.apps.sms.mixin import (CommCareMobileContactMixin, apply_leniency,
```
Comment (dimagimon) [inline corehq/apps/sms/resources/v0_5.py]: F401 'CommCareMobileContactMixin' imported but unused
F401 'InvalidFormatException' imported but unused

File: corehq/apps/sms/tests/test_registration.py [hunk 2]
```diff
@@ -56,3 +116,476 @@ def test_sms_registration(self):
...
+            self.assertEqual(response.content, '{"sms_user_registration": {"app_id": "This field is required"}}')
+
+            response = self.make_api_post(
+                self.domain1,
+                'admin@reg-api-test-1',
+                'admin@reg-api-test-1-password',
+                {'app_id': 123},
+            )
+            self.assertEqual(response.status_code, 400)
+            self.assertEqual(response.content, '{"sms_user_registration": {"app_id": "Expected type: basestring"}}')
```
Comment (dimagimon) [inline corehq/apps/sms/tests/test_registration.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/sms/tests/test_registration.py [hunk 3]
```diff
@@ -56,3 +116,476 @@ def test_sms_registration(self):
...
+            self.assertEqual(response.content, '{"sms_user_registration": {"users": "This field is required"}}')
+
+            response = self.make_api_post(
+                self.domain1,
+                'admin@reg-api-test-1',
+                'admin@reg-api-test-1-password',
+                {'app_id': '123', 'users': [{}]},
+            )
+            self.assertEqual(response.status_code, 400)
+            self.assertEqual(response.content, '{"sms_user_registration": {"phone_number": "This field is required"}}')
```
Comment (dimagimon) [inline corehq/apps/sms/tests/test_registration.py]: E501 line too long (119 > 115 characters)

**Comment by gcapalbo** [comment]
@dannyroberts

**Comment by dannyroberts** [comment]
Weird. This error must be unrelated?

```
======================================================================
FAIL: test_android_registration_from_invite (corehq.apps.sms.tests.test_registration.RegistrationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/sms/tests/test_registration.py", line 223, in test_android_registration_from_invite
    '[commcare app - do not delete] {}'.format(DUMMY_APP_INFO_URL),
  File "/mnt/commcare-hq-ro/corehq/apps/sms/tests/test_registration.py", line 88, in assertLastOutgoingSMS
    self.assertEqual(actual_texts, set(expected_texts))
AssertionError: Items in the first set but not the second:
u'Welcome to CommCareHQ! What type of phone are you using? Reply 1 for Android, 2 for other.'
Items in the second set but not the first:
'[commcare app - do not delete] http://localhost/appinfo'
```

**Comment by gcapalbo** [comment]
thanks, it is related, i'm looking into it now

**Comment by gcapalbo** [comment]
:white_check_mark:

**Pull request merged**